### PR TITLE
feat(search): Search enhancements

### LINKED ISSUE #1200: PEN-42 Notification not announced [Search] + Search Rework
## User Stories:
As a user of search, I want better loading experience while search are loading.
As a site builder, I want to supply a Multilingual search to the site users.
As a site builder, I want have an easier way to exclude components/content from site end users.
As a user of search, I want typeahead results to be accessible.
so that we provide a better search to end users.

## Acceptance Criteria
Existing ACS from another ticket with completed dev work:
1. Search results while loading present with spinner (REMOVED -- note 12/28)
1. Search should provide easy to use mechanism for site builder to exclude components/content from search results https://jira.jnj.com/browse/AESQ-3750
1. Supports Multilingual searches when there are multiple languages on a site. When a user is on a specific language then the search only returns results for that language. **(Backbone of this setup but only support single language in bodiless test site and will do a separate ticket to fully test Multilingual and activate on brand site. -- note 1/5/2022)**

Important the above 3 ACS are dev complete and in this WIP: https://github.com/johnsonandjohnson/Bodiless-JS/pull/1050 / (This story was deprioritized from work while we were trying to put accessibility into search. We would like to incorporate this work in as-is and test and build off of for AC #4. If any gaps/issues are found in the above ACs of completed work, we will consider LOE to fix and may remove or just gap.

NEW: 
 4. The "TypeAhead Suggestions" is announced to the screen reader user without interrupting the reading flow.

### PR DESCRIPTION
**Pull request by lucas-varela**
## Changes

This PR continues @dewen's work on enhancing search components.

## Related Issues

Closes #1200

### CODE REVIEW

**Review by CarlosTeixeiraCIANDT — APPROVED** [review]

**Review by hvanyo — DISMISSED** [review]

File: examples/test-site/.env.site [hunk 1]
```diff
@@ -58,14 +58,17 @@ ROBOTSTXT_SITEMAP='https://www.example.com/sitemap.xml'
...
+BODILESS_SEARCH_CONFIG='search.config.json'
 BODILESS_SEARCH_PAGE='search'
 BODILESS_SEARCH_EXPIRES='86400'
 BODILESS_SEARCH_SOURCE_PATH='./public'
 BODILESS_SEARCH_SOURCE_TYPE='html|htm'
-BODILESS_SEARCH_INDEX_URL='/lunr.idx'
+BODILESS_SEARCH_INDEX_URL='/lunr.en.json'
 BODILESS_SEARCH_INDEX_PATH='./public/lunr.idx'
 BODILESS_SEARCH_INDEX_SELECTOR='body *'
 BODILESS_SEARCH_INDEX_EXCLUDE_SELECTOR='script,noscript,style,.bg-gray-200'
```
Comment (hvanyo) [inline examples/test-site/.env.site:70]: Should be able to removed as its now in search.config.json

File: examples/test-site/.env.site [hunk 2]
```diff
@@ -58,14 +58,17 @@ ROBOTSTXT_SITEMAP='https://www.example.com/sitemap.xml'
 # - printing of design keys.
 # BODILESS_SHOWDESIGNKEYS=1
 
+BODILESS_SEARCH_ENGINE='lunr'
+BODILESS_SEARCH_CONFIG='search.config.json'
 BODILESS_SEARCH_PAGE='search'
 BODILESS_SEARCH_EXPIRES='86400'
 BODILESS_SEARCH_SOURCE_PATH='./public'
 BODILESS_SEARCH_SOURCE_TYPE='html|htm'
```
Comment (hvanyo) [inline examples/test-site/.env.site:66]: Should be able to removed as its now in search.config.json

File: examples/test-site/.env.site [hunk 3]
```diff
@@ -58,14 +58,17 @@ ROBOTSTXT_SITEMAP='https://www.example.com/sitemap.xml'
...
-BODILESS_SEARCH_INDEX_URL='/lunr.idx'
+BODILESS_SEARCH_INDEX_URL='/lunr.en.json'
 BODILESS_SEARCH_INDEX_PATH='./public/lunr.idx'
 BODILESS_SEARCH_INDEX_SELECTOR='body *'
```
Comment (hvanyo) [inline examples/test-site/.env.site:69]: Should be able to removed as its now in search.config.json

File: examples/test-site/.env.site [hunk 4]
```diff
@@ -58,14 +58,17 @@ ROBOTSTXT_SITEMAP='https://www.example.com/sitemap.xml'
 # - printing of design keys.
 # BODILESS_SHOWDESIGNKEYS=1
 
+BODILESS_SEARCH_ENGINE='lunr'
+BODILESS_SEARCH_CONFIG='search.config.json'
 BODILESS_SEARCH_PAGE='search'
```
Comment (hvanyo) [inline examples/test-site/.env.site:63]: BODILESS_SEARCH_PAGE, BODILESS_SEARCH_INDEX_URL, BODILESS_SEARCH_SOURCE_PATH probably can all be removed as well as they are in search.config.json
     "indexFileName": "lunr.en.json",
      "indexFilePath": "public",
      "searchPath": "search"

**Review by CarlosTeixeiraCIANDT — DISMISSED** [review]

File: packages/bodiless-search/src/types.ts [hunk 5]
```diff
@@ -31,13 +33,51 @@ export interface SearchEngineInterface {
...
+/**
+ * Type describes a language index setting.
+ *
+ * @param name            - language name, i.e. "English", "French" etc.
+ * @param code            - 2 letter language code (ISO 639-1).
+ * @param sourcePaths     - set of source content path for indexing. i.e. ['public', 'public/fr']
+ * @param excludePaths    - set of content paths to be excluded from indexing.
+ * @param indexFileName   - index file name for this language, used for both
+ *                          frontend and backend search process. i.e. 'lunr.json'.
+ * @param indexUrlPath    - index file url path for frontend access, i.e. '/fr'
```
Comment (misterjones) [inline packages/bodiless-search/src/types.ts:48]: Should this param be `indexUrlName`?

**Review by hvanyo — APPROVED** [review]

File: packages/bodiless-search/src/components/Search.tsx [hunk 6]
```diff
@@ -171,13 +187,37 @@ const SearchResultBase: FC<SearchResultProps> = ({
         }
       </SearchResultList>
     </SearchResultWrapper>
-  ), [searchTerm, results]);
+  );
 };
 
-const SearchBoxBase: FC<SearchProps> = ({ components, ...props }) => {
+const defaultSuggetionTotalMessage = '%total% search suggestions.';
```
[the remark this answers is not in the document]
Comment (hvanyo) [reply packages/bodiless-search/src/components/Search.tsx:193]: Approved but if we can fix const var name that would be great!

**Review by Guilherme-Almeida-Zeni — DISMISSED** [review]

**Review by hvanyo — APPROVED** [review]

**Review by TatsianaMiraniuk — APPROVED** [review]

**Pull request merged**